### PR TITLE
Implementation of NSDateComponents.copy(with: zone)

### PR DESCRIPTION
**Pull request by mamabusi**

### CODE REVIEW

File: Foundation/NSCalendar.swift [hunk 1]
```diff
@@ -1425,7 +1425,11 @@ open class NSDateComponents : NSObject, NSCopying, NSSecureCoding {
     }
     
     open func copy(with zone: NSZone? = nil) -> Any {
-        NSUnimplemented()
+        let copy = NSDateComponents()
+        copy._values = self._values
+        copy._calendar = self._calendar
+        copy._timeZone = self._timeZone
+        return copy
```
Comment (phausler) [inline Foundation/NSCalendar.swift]: this won't really work for any subclassers right?
```swift
let newObj = NSDateComponents()
newObj.calendar = calendar
newObj.timeZone = timeZone
newObj.era = era
newObj.year = year
newObj.quarter = quarter
newObj.month = month
newObj.day = day
newObj.hour = hour
newObj.minute = minute
newObj.second = second
newObj.nanosecond = nanosecond
newObj.week = week
newObj.weekOfYear = weekOfYear
newObj.weekOfMonth = weekOfMonth
newObj.yearForWeekOfYear = yearForWeekOfYear
newObj.weekday = weekday
newObj.weekdayOrdinal = weekdayOrdinal
if isLeapMonthSet {
    newObj.leapMonth = leapMonth
}
return newObj
```

Comment (mamabusi) [inline Foundation/NSCalendar.swift]: @phausler Thank you for looking into this.

Wrote a code snippet to subclass NSDateComponents:
```
class Test: NSDateComponents { 
 }
 
func test_subclass() { 
        let obj = Test()
        obj.year = 1987
        obj.month = 3
        obj.day = 17
        obj.hour = 14
        obj.minute = 20
        obj.second = 0
        let newObj = obj.copy(with: nil) as! NSDateComponents
        print(newObj.year)
        print(newObj.month)
        print(newObj.day)
    }
```
Executing the above prints the year month and day as expected. 

Are there shortcomings for subclassers that I am missing?

Thanks.

**Comment by mamabusi** [comment]
@phausler Made changes to the implementation addressing the review comments. Please have a look. Thanks!

**Comment by phausler** [comment]
@swift-ci Please test

**Pull request merged**